engineering/standards/frontend.md

```diff
@@ -326,6 +326,104 @@ TypeScript types (including interfaces) which describe API requests and response
+    ```
+
+- ❌ It's not okay to use events on our components
+
+    ```svelte
+    <Button on:click={handleClick}>
+    ```
+
+- ✅ **Our components should use props instead of events**
+
+    ```svelte
+    <Button onClick={handleClick}>
+    ```
+
+- When consuming third party components that use events, it's okay to listen to those events directly. (No need to wrap it in a component that complies with our code standards.)
+
+- For example, in a `Child.svelte` component:
+
+    - ❌ Events are dispatched as follows. Don't do this.
+
+        ```svelte
+        <script lang="ts">
+          import { createEventDispatcher } from 'svelte';
+
+          const dispatch = createEventDispatcher<{ valueChange: number }>();
+
+          dispatch('valueChange', 42);
+        </script>
+        ```
+
+    - ✅ Use props instead.
+
+        ```svelte
+        <script lang="ts">
+          export onValueChange: (value: number) => void;
+
+          onValueChange(42);
+        </script>
+        ```
+
+- Then, in the consuming (parent) component:
+
+    - ❌ Events would be handled as follows:
+
+        ```svelte
+        <script lang="ts">
+          import Child from `./Child.svelte`;
+
+          function handleValueChange({ detail: value }: CustomEvent<number>) {
+                                  // ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Yuck!
+            console.log(`Child reports that its value is: ${value}`)
+          }
+        </script>
+
+        <Child on:valueChange={handleValueChange}/>
+        ```
+
+    - ✅ With props:
+
+        ```svelte
+        <script lang="ts">
+          import Child from `./Child.svelte`;
+
+          function handleValueChange(value: number) {
+                                  // ^^^^^^^^^^^^^ Nicer!
+            console.log(`Child reports that its value is: ${value}`)
+          }
+        </script>
+
+        <Child onValueChange={handleValueChange}/>
+        ```
+
+- Rationale
+
+    | | Events | Props |
+    | --  | -- | -- |
+    | Synchronicity | ❌ Events are always synchronous. There's no way to await the `dispatch` call. | ✅ Props give us the ability to define asynchronous callbacks and then await them within the component. We do this within `SpinnerButton` for example. |
+    | Dynamic passing | ❌ Event callbacks [cannot be passed dynamically](https://stackoverflow.com/questions/60934557/how-to-bind-events-dynamically-in-svelte) to dynamically rendered components with `<svelte:component>`. This makes it impossible to do things like perform imperative logic when we're using a `ComponentAndProps` object. | ✅ Props callbacks are easily passed via spreading `{...props}` |
+    | Forwarding | ❌ Event callbacks [cannot be forwarded en masse to child components](https://github.com/sveltejs/svelte/issues/2837). Instead, each event must be manually forwarded. This makes composition more arduous. | ✅ Props callbacks are easily forwarded en masse to child components via `{...$$restProps}` |
```

Comment:
`$$restProps` requires the types being defined in all the parents inorder to ensure type safety. One could argue that, that makes composition more arduous too.